code/modules/wiremod/component.dm

```diff
@@ -0,0 +1,146 @@
+/**
+ * # Integrated Circuit Component
+ *
+ * A component that performs a function when given an input
+ *
+ * Can be attached to an integrated circuitboard, where it can then
+ * be connected between other components to provide an output or to receive
+ * an input. This is the base type of all components
+ */
+/obj/item/circuit_component
+	name = COMPONENT_DEFAULT_NAME
+	icon = 'icons/obj/module.dmi'
+	icon_state = "circuit_map"
+	inhand_icon_state = "electronic"
+
+	/// The name of the component shown on the UI
+	var/display_name = "Generic"
+
+	/// The integrated_circuit that this component is attached to.
+	var/obj/item/integrated_circuit/parent
+
+	/// A list that contains the outpurt ports on this component
+	/// Used to connect between the ports
+	var/list/datum/port/output/output_ports = list()
+
+	/// A list that contains the components the input ports on this component
+	/// Used to connect between the ports
+	var/list/datum/port/input/input_ports = list()
+
+	/// Used to determine the x position of the component within the UI
+	var/rel_x = 0
+	/// Used to determine the y position of the component within the UI
+	var/rel_y = 0
+
+	/// The power usage whenever this component receives an input
+	var/power_usage_per_input = 1
+
+	/// The current selected option
+	var/current_option
+	/// The options that this component can take on. Limited to strings
+	var/list/options
+
+	// Whether the component is removable or not. Only affects user UI
+	var/removable = TRUE
+
+/obj/item/circuit_component/Initialize()
+	. = ..()
+	if(name == COMPONENT_DEFAULT_NAME)
+		name = "[lowertext(display_name)] [COMPONENT_DEFAULT_NAME]"
+	if(length(options))
+		current_option = options[1]
+
+/obj/item/circuit_component/Destroy()
+	if(parent)
+		parent.remove_component(src)
```

Comment:
Put `parent = null` after this. Even if this is guaranteed by `remove_component`, it's still nice to be extra careful.